drivers/d3d12/rendering_device_driver_d3d12.cpp

```diff
@@ -6094,6 +6099,7 @@ Error RenderingDeviceDriverD3D12::_check_capabilities() {
 				vrs_capabilities.primitive_in_multiviewport = options6.PerPrimitiveShadingRateSupportedWithViewportIndexing;
 				vrs_capabilities.ss_image_supported = true;
 				vrs_capabilities.ss_image_tile_size = options6.ShadingRateImageTileSize;
+				vrs_capabilities.ss_max_fragment_size = 8; // TODO figure out if this is supplied and/or needed
```

Comment:
@RandomShaper I've just hardcoded this to 8, we need to have a chat on how VRS is actually setup for DX12 and what sort of density map it expects. Combined with Darios remarks I have some idea to clean this up a little more.

Reply:
See also my private message on rocket chat with something I can't share publicly :)

Reply:
I've not been able to find any way of using a density map with Direct3D 12. Only a tile-based shading rate image is supported. I think the `vrs_capabilities` structure could be upgraded with separate mentions to the classical VRS image and the new density map, both with flags advertising support, or lack thereof, as well as parameters. In D3D12, all the new members would be `false`/`0`.

Reply:
@RandomShaper ok the way I made this work is that we supply a density map but we convert this internally to a shading rate image if that is what is supported (which is currently what we support for VRS). This is what `vrs.glsl` does.

This variable steers whether our maximum fragment size is a 4x4 or 8x8 (which differs per GPU)